app/uk/gov/hmrc/cataloguefrontend/service/CostEstimationService.scala

```diff
@@ -61,6 +91,110 @@ object CostEstimationService {
       Json.reads[DeploymentConfig]
   }
 
+  final case class CostedResourceUsage(
+    date: Instant,
+    serviceName: String,
+    environment: Environment,
+    totalSlots: Int,
+    yearlyCostGbp: Double
+  )
+
+  object CostedResourceUsage {
+
+    def fromResourceUsage(resourceUsage: ResourceUsage, slotCostPerYear: Double): CostedResourceUsage = {
+      val totalSlots =
+        resourceUsage.slots * resourceUsage.instances
+
+      val yearlyCostGbp =
+        totalSlots * slotCostPerYear
+
+      CostedResourceUsage(
+        resourceUsage.date,
+        resourceUsage.serviceName,
+        resourceUsage.environment,
+        totalSlots,
+        yearlyCostGbp
+      )
+    }
+
+    def toChartDataTable(costedResourceUsages: List[CostedResourceUsage]): ChartDataTable = {
+        val environments =
+          costedResourceUsages
+            .map(_.environment)
+            .distinct
+            .sorted
+
+        val resourceUsageByDateAndEnvironment =
+          costedResourceUsages
+            .groupBy(_.date)
+            .mapValues(rus => rus.map(ru => ru.environment -> ru).toMap)
+
+        val headerRow =
+          "'Date'" +: environments.map(e => s"'${e.displayString}'")
+
+        val dataRows =
+          resourceUsageByDateAndEnvironment
+            .toList
+            .sortBy(_._1)
+            .map { case (date, byEnv) =>
+              s"new Date(${date.toEpochMilli})" +:
+                environments.map(e => s"${byEnv.get(e).map(_.yearlyCostGbp).getOrElse("null")}")
+            }
+
+        ChartDataTable(headerRow +: dataRows)
+    }
+  }
+
+  final case class CostedResourceUsageTotal(
+    date: Instant,
+    totalSlots: Int,
+    yearlyCostGbp: Double
+  ) {
+    def +(other: CostedResourceUsageTotal): CostedResourceUsageTotal =
+      CostedResourceUsageTotal(date, totalSlots + other.totalSlots, yearlyCostGbp + other.yearlyCostGbp)
```

Comment:
What if the dates are not the same? Should this bomb?Would it be better to split the Date out of the "monoid"? (A value object for slots/yearlyCost seems worthwhile?)

Reply:
Yeah, fair, I think I'll just move `+` to the local scope of where it's used (which is safe, as we group by `date` there and just smash them all together... semigroupally)